akka-contrib/src/main/scala/akka/contrib/pattern/DistributedPubSubMediator.scala

```diff
@@ -548,6 +574,10 @@ class DistributedPubSubMediator(
       content = bucket.content + (key -> ValueHolder(v, valueOption))))
   }
 
+  def getCurrentTopics(): Map[String, ActorRef] = {
+    context.children.map(ref ⇒ (URLDecoder.decode(ref.path.name, "utf-8"), ref)).toMap
+  }
```

Comment:
Thanks Pattrick

I put in your code suggestions (now that I understand them, not sure how to
improve on them).  The code compiles, but I now get the following error
when attempting to run the multi-jvm tests (not just for this test but all
multi-jvm tests in akka-contrib)

Exception in thread "ScalaTest-main" java.lang.NoSuchMethodError:
scala.collection.immutable.$colon$colon.hd$1()Ljava/lang/Object

at org.scalatest.tools.Runner$.argTooShort$1(Runner.scala:1515)

[JVM-2] at
org.scalatest.tools.Runner$.parseReporterArgsIntoConfigurations(Runner.scala:1532)

[JVM-2] at
org.scalatest.tools.Runner$.runOptionallyWithPassFailReporter(Runner.scala:923)

[JVM-2] at org.scalatest.tools.Runner$.main(Runner.scala:860)

[JVM-2] at org.scalatest.tools.Runner.main(Runner.scala)

[JVM-3] Exception in thread "ScalaTest-main" java.lang.NoSuchMethodError:
scala.collection.immutable.$colon$colon.hd$1()Ljava/lang/Object;

[JVM-3] at org.scalatest.tools.Runner$.argTooShort$1(Runner.scala:1515)

[JVM-3] at
org.scalatest.tools.Runner$.parseReporterArgsIntoConfigurations(Runner.scala:1532)

Can't get past this or even figure out what is behind this.  Any ideas?



On Sat, Feb 21, 2015 at 3:48 AM, Patrik Nordwall <notifications@github.com>
wrote:

> In
> akka-contrib/src/main/scala/akka/contrib/pattern/DistributedPubSubMediator.scala
> <https://github.com/akka/akka/pull/16324#discussion_r25120844>:
>
> > @@ -548,6 +574,10 @@ class DistributedPubSubMediator(
> >        content = bucket.content + (key -> ValueHolder(v, valueOption))))
> >    }
> >
> > +  def getCurrentTopics(): Map[String, ActorRef] = {
> > +    context.children.map(ref ⇒ (URLDecoder.decode(ref.path.name, "utf-8"), ref)).toMap
> > +  }
>
> The registry contains one bucket per node, we iterate all of those.
>
> Each bucket contains ActorRef of the actors that have been registered with
> Put and also child ActorRef for topics that have been registered with
> Subscribe. We are only interested in the latter here. Those are are keyed
> by their path without address, i.e. the subscribe topics all start with the
> same string as the self (DistributedPubSubMediator actor) path. That is
> why we filter out only topics by looking at the prefix.
>
> Then there is a special kind of topics for the group feature. We exclude
> those also (!topic.contains('/') // exclude group topics).
>
> Remaining should be all topics that have been registered with Subscribe.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/akka/akka/pull/16324/files#r25120844>.
>